src/mame/nabu/nabupc.cpp

```diff
@@ -0,0 +1,509 @@
+//  UTILITY FUNCTIONS
+//**************************************************************************
+
+// Bit Manipulation
+namespace {
+	template<typename T> constexpr T BIT_SHIFT(unsigned n, unsigned v)
+	{
+		return (T)v << n;
+	}
+
+	template<typename T> void BIT_SET(T& w , unsigned n, unsigned v)
+	{
+		T mask = ~BIT_SHIFT<T>(n, 1);
+		w = (w & mask) | BIT_SHIFT<T>(n, v);
+	}
+}
+
+// Priority Encoder
+/** @brief F9318 input lines */
+typedef enum {
+	PRIO_IN_EI = (1<<8),
+	PRIO_IN_I7 = (1<<7),
+	PRIO_IN_I6 = (1<<6),
+	PRIO_IN_I5 = (1<<5),
+	PRIO_IN_I4 = (1<<4),
+	PRIO_IN_I3 = (1<<3),
+	PRIO_IN_I2 = (1<<2),
+	PRIO_IN_I1 = (1<<1),
+	PRIO_IN_I0 = (1<<0),
+	/* masks */
+	PRIO_I7 = PRIO_IN_I7,
+	PRIO_I6_I7 = (PRIO_IN_I6 | PRIO_IN_I7),
+	PRIO_I5_I7 = (PRIO_IN_I5 | PRIO_I6_I7),
+	PRIO_I4_I7 = (PRIO_IN_I4 | PRIO_I5_I7),
+	PRIO_I3_I7 = (PRIO_IN_I3 | PRIO_I4_I7),
+	PRIO_I2_I7 = (PRIO_IN_I2 | PRIO_I3_I7),
+	PRIO_I1_I7 = (PRIO_IN_I1 | PRIO_I2_I7),
+	PRIO_I0_I7 = (PRIO_IN_I0 | PRIO_I1_I7),
+}   f9318_in_t;
+
+/** @brief F9318 output lines */
+typedef enum {
+	PRIO_OUT_Q0 = (1<<0),
+	PRIO_OUT_Q1 = (1<<1),
+	PRIO_OUT_Q2 = (1<<2),
+	PRIO_OUT_EO = (1<<3),
+	PRIO_OUT_GS = (1<<4),
+	/* masks */
+	PRIO_OUT_QZ = (PRIO_OUT_Q0 | PRIO_OUT_Q1 | PRIO_OUT_Q2)
+}   f9318_out_t;
+
+/**
+ * @brief F9318 priority encoder 8 to 3-bit
+ *
+ * Emulation of the F9318 chip (pin compatible with 74348).
+ *
+ * <PRE>
+ *            F9318
+ *         +---+-+---+
+ *         |   +-+   |         +---------------------------------+----------------+
+ *    I4' -|1      16|-  Vcc   |              input              |     output     |
+ *         |         |         +---------------------------------+----------------+
+ *    I5' -|2      15|-  EO'   |      EI I0 I1 I2 I3 I4 I5 I6 I7 | GS Q0 Q1 Q2 EO |
+ *         |         |         +---------------------------------+----------------+
+ *    I6' -|3      14|-  GS'   | (a)  H  x  x  x  x  x  x  x  x  | H  H  H  H  H  |
+ *         |         |         | (b)  L  H  H  H  H  H  H  H  H  | H  H  H  H  L  |
+ *    I7' -|4      13|-  I3'   +---------------------------------+----------------+
+ *         |         |         | (c)  L  x  x  x  x  x  x  x  L  | L  L  L  L  H  |
+ *    EI' -|5      12|-  I2'   | (d)  L  x  x  x  x  x  x  L  H  | L  H  L  L  H  |
+ *         |         |         | (e)  L  x  x  x  x  x  L  H  H  | L  L  H  L  H  |
+ *    Q2' -|6      11|-  I1'   | (f)  L  x  x  x  x  L  H  H  H  | L  H  H  L  H  |
+ *         |         |         | (g)  L  x  x  x  L  H  H  H  H  | L  L  L  H  H  |
+ *    Q1' -|7      10|-  I0'   | (h)  L  x  x  L  H  H  H  H  H  | L  H  L  H  H  |
+ *         |         |         | (i)  L  x  L  H  H  H  H  H  H  | L  L  H  H  H  |
+ *   GND  -|8       9|-  Q0'   | (j)  L  L  H  H  H  H  H  H  H  | L  H  H  H  H  |
+ *         |         |         +---------------------------------+----------------+
+ *         +---------+
+ * </PRE>
+ */
+static __inline f9318_out_t f9318(f9318_in_t in)
```

Comment:
Please don’t use `__inline` - use C++ `inline`, or use `ATTR_FORCE_INLINE` from osdcomn.h for things that absolutely must be inlined.  Also, you can throw all the local stuff in an anonymous namespace and not have to put `static` everywhere.